### PR TITLE
Deserialize Node View IsSetAsInput property on DynamoMode File Open

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

[QNTM-3876](https://jira.autodesk.com/browse/QNTM-3876)
As a Dynamo Dev, I want to deserialize Node View IsSetAsInput property on DynamoMode File Open

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @saintentropy @ramramps 

### FYIs

@BogdanZavu @dobriai

### CODE REVIEW

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 1]
```diff
@@ -350,7 +350,8 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             // nodes
             var nodes = obj["Nodes"].ToObject<IEnumerable<NodeModel>>(serializer);
 
-            // nodes
+            // Setting Inputs
+            // Required in UI less mode by Dynamo Player that NodeModel.Name and NodeModel.IsSetAsInput are set
```
Comment (ramramps) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: headless

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 2]
```diff
@@ -56,8 +56,8 @@ public class ExtraNodeViewInfo
         public double Y;
         public bool ShowGeometry;
         public bool Excluded;
-        public bool IsInput;
-        public bool IsOutput;
+        public bool IsSetAsInput;
+        public bool IsSetAsOutput;
```
Comment (ramramps) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: API change?

**Comment by ramramps** [comment]
@QilongTang Is the API changed here?

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs (hunk 2 above)
Comment (QilongTang) [reply src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: @ramramps This chass is introduced as part of Dynamo 2.0

Comment (ramramps) [reply src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: so, if we change this, we should let Alias team know this?

**Comment by mjkkirschner** [comment]
@QilongTang  - @saintentropy brought up a good point... if we do this we should also transfer other properties like frozen and maybe preview - otherwise graphs will run which users do not expect to run which will modify revit files and be really difficult to debug.

But if we start doing this, we will have a dependency on the model properties existing and so will never remove them from the model....

What this points to is that the placement of some of these properties on the view may have been a mistake. 😢 

It's okay with me to do this transfer, but we should be aware of the consequences to API and dependencies going forward.

**Comment by QilongTang** [comment]
@ramramps I don't think it counts before 2.0 went out since the class touches does not exist in RC1.3.3. Also there will be refinary changes depending on this PR but @saintentropy metioned should be minimal. We did discuss if `Excluded` need to be set as well since it actually will change graph execution result, but there is workaround if a particular client needs it in RC2.0.1. What we should do is making the schema supporting selection and drop down in **Inputs**, and find a way to deal with Node Name

**Comment by mjkkirschner** [comment]
But player does need it - don't they? - if player does not create a view model they will run all nodes in the graph unless we transfer that property here. What is the workaround?

**Comment by QilongTang** [comment]
@mjkkirschner Currently I don't know if they need it, in my imagination they will need it but need to hear from @BogdanZavu @dobriai, the workaround is to add **exclued** parsing as well in 2.0.1

**Comment by mjkkirschner** [comment]
I just want to make one more point - this and the other PR (getting name from the view) have introduced inconsistency, some properties are transfered, some are not, even though those properties now are correct at load time, they will be wrong when saved.

If we make these changes I think they need to be documented for API developers, in addition our original constraints (requiring viewModels for view properties also should get some documentation since they non obvious)

**Comment by QilongTang** [comment]
@mjkkirschner Exactly, but player is not capable of saving node **name** or changing the **isSetAsInput** flag runtime, it almost become our invisible requirements for now and I am open for suggestions where we should document this fact

**Comment by mjkkirschner** [comment]
consider an extension developer, or a CLI user instead then.

**Comment by ramramps** [comment]
Can''t we create an extension method that takes the property name and returns the value? To be more generic instead of very specific? @QilongTang

**Comment by QilongTang** [comment]
@ramramps That means we need to parse the Json at least one more time in that utility method for very minimal gain in my mind so I did not do that

**Comment by ramramps** [comment]
like mike said, what if we need more parameters from view?

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs (hunk 2 above)
Comment (QilongTang) [reply src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: Sure, although they most likely will not be using this public class

Comment (ramramps) [reply src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: I think so.. but we said if we changed the API, we will let them know of the new binaries..

**Comment by ramramps** [comment]
I still think, we should speed up the ViewModel creation... 
but I am ok if this approach supports player for now...
but are we saying others to follow this approach to get any of the view properties?

**Comment by saintentropy** [comment]
I think my point is that because these properties exist on the model, they are not set properly until a view is initialized.  It is unclear to user of the API who are using dynamo core in the headless state (right now, Player, Black Team, and soon Purple team) that these values are not correct and in the case of freeze that the graph will execute incorrectly.  Documentation can fix this but it feels like a regression in the case of the CLI.  The solution here feels like a bandaid but perhaps a necessary one

**Comment by saintentropy** [comment]
I did confirm that the player ignores freeze state.  Not good in a Revit context :-(

**Comment by dobriai** [comment]
I am not sure that it means for the DynamoPlayer to "ignore" frozen state. DP only cares about displaying and setting values of input nodes. So, if an Input Node is frozen, then DP, maybe incorrectly, displays it and allows its value to be set - OK. But then DP does not control explicitly what Nodes get executed or not, at least AFAIK. So, it is up to the Model to decide what nodes to execute or ignore, in the end.
...or I totally don't get it :-)

**Comment by ramramps** [comment]
@saintentropy what is meant by ignore here? does the frozen node executes or nodes allows the freeze property to be set even though it is actually frozen?

**Comment by saintentropy** [comment]
Sorry I am confusing the issue of initializing the views for inputs for player with what happens when a user presses play.  Currently dynamo player only loads the model for the graph for both gathering inputs and headles execution.  If a user has a player graph with frozen nodes these will be executed as un-frozen by player unless we also load those states similar to the isInput and name as suggested in his PR

**Comment by saintentropy** [comment]
@ramramps does that answer you question?

**Comment by ramramps** [comment]
yes thank @saintentropy... 
I am wondering what other use case we are missing here? 
@QilongTang @mjkkirschner I think we should talk on monday , and get things straight. Thoughts?

**Comment by mjkkirschner** [comment]
yes, we should consider reverting the other Name PR and discuss this throughly before moving forward.

**Comment by BogdanZavu** [comment]
@mjkkirschner @QilongTang I would cast a vote in favor of transferring back some properties (which are very few and simple so far at least) rather than speed up the view model (which of course would still be very nice to have). As an API dev client I really liked the whole separation between model and view. I think this conceptual separation and the ability to run the model without the view is something that all products should have especially now in the cloud era.
I'm a little worried about what @mjkkirschner said about executing frozen nodes when we don't have the view model but I hope that's something that can be easily fix/avoid.

**Comment by Racel** [comment]
@QilongTang @saintentropy @ramramps @mjkkirschner @kronz - After discussion with the Revit team, we are in agreement that Player in Dynamo 2.0 does not need to support frozen nodes or background geometry preview. There are currently no player use cases or issues brought up by Player customers where these properties are needed at this current time. Also, it is not a regression from 1.3.x of Player functionality. We understand that this may be a useful functionality to have later for other third party developers, so we plan on fixing this in a later 2.x release as this does not affect the API. But, for the upcoming 2.0 release, we are going to merge this fix right now and begin testing.

@QilongTang for the follow-on work, can we please file a task to be able to support frozen and background geometry preview in a headless state?

FYI @BogdanZavu @dobriai

**Comment by QilongTang** [comment]
@ramramps @mjkkirschner @saintentropy @kronz @Racel @BogdanZavu @dobriai https://jira.autodesk.com/browse/QNTM-3892 Created for part of Dynamo 2.1 Epic. Will merge this as it is and cherry-pick to meet Player's current requirement first

**Comment by saintentropy** [comment]
@Racel Can you explain more about the comment that this is not a regression.  Previously frozen nodes did not execute in 1.3.x when you used player.  Now they do?

**Comment by Racel** [comment]
@saintentropy - "Frozen nodes not working" is not a regression. They will continue to not work in Dynamo 2.0

**Comment by mjkkirschner** [comment]
But it is a regression... they worked in Dynamo player 1.x....

**Comment by mjkkirschner** [comment]
or it's a change in behavior by design ? :)

**Comment by saintentropy** [comment]
In CLI and Player, Frozen nodes were Frozen when you executed them in 1.x

**Comment by QilongTang** [comment]
@saintentropy @mjkkirschner Dynamo Player team will test and confirm that, in the same time, both team will blog post that the Dynamo Player come with RC2.0.0 is a Dynamo Player Preview Release so there can be areas not working. The official Dynamo Player release come with R2020

**Comment by mjkkirschner** [comment]
I am fine with making a decision about what functionality works in a new release and communicating it, but it's clear this is a change in behavior from previous versions, and we also know that users were using this functionality in revit graphs, which player runs. I can see the argument that users do not need this functionality in graphs that will be solely run in CLI or player though.

**Comment by Racel** [comment]
@mjkkirschner Are you saying that frozen is just broken even when you open the graph in dynamo? Or just not accessible headlessly?

**Comment by mjkkirschner** [comment]
I am saying that in 2.0 - the state of frozen nodes is not set when a graph is run headless.

In 1.3 - nodes that had been set to frozen, when run headless, would have that state set correctly. Now they will not. I am okay with that - but to users it will appear as a regression.

**Comment by Racel** [comment]
@dobriai @BogdanZavu Please note the comment from Mike above. We want to make sure we are on the same page here. So, apparently there is a regression on the frozen nodes. Are we ok with this?

**Comment by BogdanZavu** [comment]
@Racel We discussed the issue and we think is ok for now as long as it is specified in the release notes. We will need this however for the official Revit 2020 ( I suspect it's going to be fixed much sooner that that )

**Pull request merged**